src/kernel/hartlock.c

```diff
@@ -0,0 +1,30 @@
+#include "hart_locals.h"
+#include "hartlock.h"
+#include "irq.h"
+#include "panic.h"
+#include "types.h"
+
+void hart_lock(void) {
+	u32 old_irq_state = get_irq_state();
+	irq_disable();
+	struct hart_locals *hart = get_hart_locals();
+	if (hart->noff == 0) {
+		hart->intena = old_irq_state;
+	}
+	hart->noff += 1;
+}
+
+void hart_unlock(void) {
+	if (irq_get_state() != 0) {
+		panic("hart_unlock: interruptable");
+	}
+	struct hart_locals *hart = get_hart_locals();
+	if (hart->noff < 1) {
+		panic("hart_unlock: noff < 1");
```

Comment:
ideally, error messages should let the programmer know what they did wrong even if they don't know the implementation details of the function giving the error

e.g. `hart_unlock called without a matching hart_lock`

if the error ought to be impossible without the user messing with the private state of the module somehow (the noff one is, right?), you can use `assert` -- that'll include the condition that failed already